core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -643,7 +643,7 @@ public ShareAcquiredRecords acquire(
                         // acquire subset of offsets from the in-flight batch but only if the
                         // complete batch is available yet. Hence, do a pre-check to avoid exploding
                         // the in-flight offset tracking unnecessarily.
-                        if (inFlightBatch.batchState() != RecordState.AVAILABLE) {
+                        if (inFlightBatch.batchState() != RecordState.AVAILABLE || inFlightBatch.batchRollbackState() != null) {
```

Comment:
Now, actually, `batchRollbackState()` will throw an exception if the rollback state is null. That's probably not what you mean here. Maybe `hashBatchRollbackState()` that doesn't throw an exception would be better.

Reply:
`batchRollbackState() `will throw an error if `batchState` is null i.e. the batch is maintained per offsets. If the rollback state of the batch is null, it will be returned as it is, it won't throw an exception.

Reply:
Sure? If we call `InFlightState.completeStateTransition`, then the rollback state will be set to null. I wonder whether there's a code path in which a null could result.

Reply:
Yeah, if the `rollbackState` is set to null, when we are trying to retrieve `rollbackState`, that case will get handled in the function `rollbackState()` [here](https://github.com/apache/kafka/pull/17965/files#diff-83f8b0eae4289cefc39e38e4b85e64d49610ceb900eb456285f1ab8d240d9555R2343). It will not throw an exception. IIUC, only case when batchRollbackState() can throw an exception is if `batchState` is null i.e. the batch is maintained per offsets.